akka-cluster-sharding/src/main/scala/akka/cluster/sharding/dynamic/DynamicShardAllocationStrategy.scala

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.dynamic
+
+import akka.actor.Actor
+import akka.actor.ActorLogging
+import akka.actor.ActorRef
+import akka.actor.ActorSystem
+import akka.actor.Address
+import akka.actor.ExtendedActorSystem
+import akka.actor.LocalActorRef
+import akka.actor.NoSerializationVerificationNeeded
+import akka.actor.Props
+import akka.actor.Stash
+import akka.cluster.Cluster
+import akka.cluster.ddata.DistributedData
+import akka.cluster.ddata.LWWMapKey
+import akka.cluster.ddata.Replicator.Changed
+import akka.cluster.ddata.Replicator.Subscribe
+import akka.cluster.sharding.ShardCoordinator
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.event.Logging
+import akka.util.Timeout
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+// TODO reduce logging
+object DynamicShardAllocationStrategy {
+
+  type ShardRegion = ActorRef
+
+  // local only messages
+  private final case class GetShardLocation(shard: ShardId) extends NoSerializationVerificationNeeded
+  private final case object GetShardLocations extends NoSerializationVerificationNeeded
+
+  private final case class GetShardLocationsResponse(desiredAllocations: Map[ShardId, ShardLocation])
+      extends NoSerializationVerificationNeeded
+  private final case class GetShardLocationResponse(address: Option[Address]) extends NoSerializationVerificationNeeded
+
+  // TODO serializer, datatype used with ddata, use a normal class to evolve
+  final case class ShardLocation(address: Address)
```

Comment:
I think we should consider using a plain `String` in the ddata Map becuase that and a few primitive types are optimized to not serialize each entry separately.

We must be able to support 1000s (maybe 10000s) shards and the size of the Map must be of reasonable in size. Maybe we need to split up in several maps, as we do for the remember entities.

Reply:
ah good point. I'll switch to strings and split it into multiple maps as you suggest